prover/src/arithmetic/add.rs

```diff
@@ -0,0 +1,145 @@
+use std::matches;
+
+use crate::{columns::*, utils::generate_inst_trace};
+use plonky2::field::types::Field;
+use vm_core::program::{instruction::*, REGISTER_NUM};
+use vm_core::trace::trace::{MemoryTraceCell, Step};
+
+use plonky2::field::extension::Extendable;
+use plonky2::field::goldilocks_field::GoldilocksField;
+use plonky2::field::packed::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+
+pub(crate) fn eval_packed_generic<P: PackedField>(
+    lv: &[P; NUM_INST_COLS],
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    // Get ADD data from trace.
+    let is_add = lv[COL_S_ADD];
+    let output = lv[COL_OP_0];
+    let input0 = lv[COL_OP_1];
+    let input1 = lv[COL_OP_2];
+
+    // TODO: We use range_check to check input/output are in 32 bits.
+    // range_check(output, 32);
+    // range_check(input0, 32);
+    // range_check(input0, 32);
+
+    // Do a local addition.
+    let unreduced_output = input0 + input1;
+    let output_diff = unreduced_output - output;
+
+    // Constraint addition.
+    yield_constr.constraint(is_add * output_diff);
+}
+
+pub(crate) fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    lv: &[ExtensionTarget<D>; NUM_INST_COLS],
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    // Get ADD data from trace.
+    let is_add = lv[COL_S_ADD];
+    let output = lv[COL_OP_0];
+    let input0 = lv[COL_OP_1];
+    let input1 = lv[COL_OP_2];
+
+    let unreduced_output = builder.add_extension(input0, input1);
```

Comment:
Why reuse the circuit of plonky2?

Reply:
for u32_add, the trace should be
a + b*1 = d
d_lo + d_hi *2^32 = d
rangecheck(d_lo)
bool(d_hi)

Reply:
New version:
for u32_add, the trace should be
a + b * 1 = d
d_lo + d_hi *2^32 = d
d_lo_lo16 + d_lo_hi_16 *2^16 = d_lo
range check(d_lo_lo16, d_lo_hi16)
bool(d_hi)